central/complianceoperator/v2/report/manager/watcher/scanconfigwatcher.go

```diff
@@ -0,0 +1,196 @@
+)
+
+// GetScanConfigFromScan returns the ScanConfiguration associated with the given scan
+func GetScanConfigFromScan(ctx context.Context, scan *storage.ComplianceOperatorScanV2, scanConfigDS scanConfigurationDS.DataStore) (*storage.ComplianceOperatorScanConfigurationV2, error) {
+	return scanConfigDS.GetScanConfigurationByName(ctx, scan.GetScanConfigName())
+}
+
+// ScanConfigWatcher determines if a ScanConfiguration has running scans or has completed
+type ScanConfigWatcher interface {
+	PushScanResults(results *ScanWatcherResults) error
+	Subscribe(snapshotID string)
+	Stop()
+	Finished() concurrency.ReadOnlySignal
+}
+
+// ScanConfigWatcherResults is returned when the watcher detects all the scans are completed
+type ScanConfigWatcherResults struct {
+	Ctx               context.Context
+	WatcherID         string
+	ReportSnapshotIDs []string
+	ScanConfig        *storage.ComplianceOperatorScanConfigurationV2
+	ScanResults       map[string]*ScanWatcherResults
+	Error             error
+}
+
+type scanConfigWatcherImpl struct {
+	ctx     context.Context
+	cancel  func()
+	scanC   chan *ScanWatcherResults
+	stopped *concurrency.Signal
+
+	scanDS    scan.DataStore
+	profileDS profileDatastore.DataStore
+
+	snapshotsLock     sync.Mutex
+	readyQueue        readyQueue[*ScanConfigWatcherResults]
+	scanConfigResults *ScanConfigWatcherResults
+	scansToWait       set.StringSet
+	totalResults      int
+}
+
+// NewScanConfigWatcher creates a new ScanConfigWatcher
+func NewScanConfigWatcher(ctx context.Context, watcherID string, sc *storage.ComplianceOperatorScanConfigurationV2, scanDS scan.DataStore, profileDS profileDatastore.DataStore, queue readyQueue[*ScanConfigWatcherResults], snapshotIDs ...string) *scanConfigWatcherImpl {
+	watcherCtx, cancel := context.WithCancel(ctx)
+	finishedSignal := concurrency.NewSignal()
+	ret := &scanConfigWatcherImpl{
+		ctx:        watcherCtx,
+		cancel:     cancel,
+		stopped:    &finishedSignal,
+		scanDS:     scanDS,
+		profileDS:  profileDS,
+		scanC:      make(chan *ScanWatcherResults),
+		readyQueue: queue,
+		scanConfigResults: &ScanConfigWatcherResults{
+			Ctx:               ctx,
+			WatcherID:         watcherID,
+			ScanConfig:        sc,
+			ReportSnapshotIDs: snapshotIDs,
+			ScanResults:       make(map[string]*ScanWatcherResults),
+		},
+		scansToWait: set.NewStringSet(),
+	}
+	timeout := time.NewTicker(defaultScanConfigTimeout)
+	go ret.run(timeout.C)
+	return ret
+}
+
+// PushScanResults queues a ScanWatcherResults to be handled
+func (w *scanConfigWatcherImpl) PushScanResults(results *ScanWatcherResults) error {
+	select {
+	case <-w.ctx.Done():
+		return errors.New("The watcher is stopped")
+	case w.scanC <- results:
+		return nil
+	}
+}
+
+// Subscribe a new readyQueue to the watcher
+func (w *scanConfigWatcherImpl) Subscribe(id string) {
+	concurrency.WithLock(&w.snapshotsLock, func() {
+		w.scanConfigResults.ReportSnapshotIDs = append(w.scanConfigResults.ReportSnapshotIDs, id)
+	})
+}
```

Comment:
I see two issues here:
1. It doesn't look like it does what says in the comment
2. This is used only in test so far, so it is hard to say how that maybe used in the wild - there is a risk that `w.scanConfigResults` may be nil when this is called.

Reply:
> It doesn't look like it does what says in the comment

The comment is wrong. I need to updated.

> This is used only in test so far, so it is hard to say how that maybe used in the wild - there is a risk that w.scanConfigResults may be nil when this is called.

This will be used in the follow up PR. It will have some changes once the datastore is introduced (also in the next PR). I'll add a check to make sure `scanConfigResults` is not nil.

Reply:
Done

Reply:
Agh yeah.. I fix (2) and not (1) 🤦

Reply:
Done